docs: guide quick start app creation using the slack cli (#2535)
slackapi/bolt-js: docs/content/getting-started.mdx

> ---

Author: 🫡 No! In fact quite the opposite. `.mdx` is not a familiar extension to me so I am glad to make this change.

FWIW the linked page is confusing to me and I don't know which markdown format is used in practice but I am hoping most for what's forgiving 🙏 

> By default, Docusaurus v3 uses the MDX format for all files (including `.md` files) for historical reasons.
> 
> `.md` files will use the CommonMark format

title: Quick start with Bolt for JavaScript
sidebar_label: Quick start
---

This quick start guide aims to help you get a Slack app using Bolt for JavaScript up and running as soon as possible!

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

When complete, you'll have a local environment configured with a customized [app](https://github.com/slack-samples/bolt-js-getting-started-app) running that responds to a simple greeting.

:::note

In search of the complete guide to building an app from scratch? Check out the [building an app](/building-an-app) guide.

:::

#### Prerequisites

A few tools are needed for the following steps. We recommend using the [**Slack CLI**](https://tools.slack.dev/slack-cli/) for the smoothest experience, but other options remain available.

> <Tabs groupId="cli-or-terminal">

Reviewer: i'm tripping ove this a bit because the cli is _in_ the terminal. is there another delineator to use?

Author: @lukegalbraithrussell This is so confusing to me too. I don't have a better suggestion at the moment since:

- `cli` is for the Slack CLI
- `terminal` is for both direct commands and the browser

IMO this difference can be understood alright in the tab titles, but that doesn't make it less confusing... 🤔

<TabItem value="cli" label="Slack CLI">

Install the latest version of the Slack CLI to get started:

- **Mac / Linux**: https://tools.slack.dev/slack-cli/guides/installing-the-slack-cli-for-mac-and-linux
- **Windows**: https://tools.slack.dev/slack-cli/guides/installing-the-slack-cli-for-windows

Then confirm a successful installation with the following command:

```sh
$ slack version
```

An authenticated login is also required if this hasn't been done before:

```sh
$ slack login
```

</TabItem>
<TabItem value="terminal" label="Terminal">

Tooling for the terminal can also be used to follow along:

- **Git**: https://git-scm.com/downloads
- **Node.js**: https://nodejs.org/en/download

Once installed, make sure recent versions are being used:

```sh
$ git --version
$ node --version
```

</TabItem>
</Tabs>

> :::tip

Reviewer: ```suggestion
:::info[A place to belong]
```

I did a lazy thing where, to get rid of the all caps default titles, i just had it strip all capitalization. so now it's just a lower case "tip" if you don't override the default title. but we like customzied titles anyway!

Author: @lukegalbraithrussell I had no idea how custom titles were done! 🤯

I'm a big fan of all of these suggestions. Thanks as always for thinking well about the kind reader 🤖 ✨

A workspace where development can happen is also needed.

> We recommend using developer [sandboxes](https://docs.slack.dev/tools/developer-sandboxes) to avoid disruptions where real work gets done.

Reviewer: ```suggestion
We recommend using [developer sandboxes](https://docs.slack.dev/tools/developer-sandboxes) to avoid disruptions where real work gets done.
```

```sh
$ slack create first-bolt-app --template slack-samples/bolt-js-getting-started-app
$ cd first-bolt-app
```

After a project is created you'll have a `package.json` file with project details and a `.slack` directory for application use.

A few other files exist too, but we'll visit these later.

</TabItem>
<TabItem value="terminal" label="Terminal">

A blank template can be cloned to start with project scaffolding:

```sh
$ git clone https://github.com/slack-samples/bolt-js-getting-started-app first-bolt-app
$ cd first-bolt-app
$ npm install
```

Outlines of a project are taking shape, so we can move on the creating an app!

</TabItem>
</Tabs>

## Creating an app {#creating-an-app}

Before you can start developing with Bolt, you will want to create a Slack app.

<Tabs groupId="cli-or-terminal">
<TabItem value="cli" label="Slack CLI">

The getting started app template contains a `manifest.json` file with details about an app that we will use to get started. Use the following command and select "Create a new app" to install the app to the team of choice:

```sh
$ slack run
...
[INFO]  bolt-app ⚡️ Bolt app is running!
```

> After the app has started, celebrate, then interrupt the process with `^C` to stop it for now.

Reviewer: <3

```

The above command works on Linux and macOS but [similar commands are available on Windows](https://superuser.com/questions/212150/how-to-set-env-variable-in-windows-cmd-line/212153#212153).

:::warning

Treat your tokens like a password and [keep it safe](https://docs.slack.dev/authentication/best-practices-for-security). Your app uses these to retrieve and send information to Slack.

:::

A bot token is also needed to interact with the Web API methods as your app's bot user. We can gather this as follows:

1. Navigate to the **OAuth & Permissions** on the left sidebar and install your app to your workspace to generate a token.
2. After authorizing the installation, you'll return to the **OAuth & Permissions** page and find a **Bot User OAuth Access Token**:

![OAuth Tokens](/img//bot-token.png "Bot OAuth Token")

3. Copy the bot token `xoxb` from the **OAuth & Permissions page** and then store it in a new environment variable:

```sh
$ export SLACK_BOT_TOKEN=xoxb-example
```

</TabItem>
</Tabs>

## Running the app

After the app has been created, it is time to run it:

<Tabs groupId="cli-or-terminal">
<TabItem value="cli" label="Slack CLI">

> ```sh

Reviewer: `sh` isn't default in docusaurus i believe. if you'd be so kind, add this to the docusaurus.config.js

```
  themeConfig: // already exists
  ...
      prism: { // already exists
        darkTheme: prismThemes.dracula, // already exists
        additionalLanguages: ['bash'] // put this here pls
      },
```

Author: @lukegalbraithrussell Nice catch! I'm surprised `bash` is not a default language but it was no problem adding this in b61e880113d870bff166a3cd30ac23fcd99759e4 📚

$ slack run
```

</TabItem>
<TabItem value="terminal" label="Browser">

```script
$ npm run start
```

</TabItem>
</Tabs>

Logs in the console will let you know that the app is up and running:

```sh
[INFO]  bolt-app ⚡️ Bolt app is running!
```

With the app running, add your app's bot to a channel or send it a direct message saying "hello" and wait for a response. Clicking the attached button will post another message to channel.

## Updating the app

> At this point, you're running the getting started Bolt for JavaScript [app](https://github.com/slackapi/bolt-js-getting-started-app)!

Reviewer: ```suggestion
At this point, you're successfully running the getting started Bolt for JavaScript [app](https://github.com/slackapi/bolt-js-getting-started-app)!
```

> The defaults included leave opportunities abound, so to personalize this app let's now update the app name then respond with a kind farewell.

Reviewer: this reads like taylor's docs voice and makes me nostalgic <3

Author: Imitation is the highest form of learning - or something somewhat similar I think - from one of the best 🤩

#### Changing the app name

> With intentions of the app's purpose known, return back to app settings and make changes that seem right.

Reviewer: ```suggestion
With intentions of the app's purpose known, return back to app settings and make changes that feel right to you.
```

- **App name**: The label describing your app. Found on the "Basic Information" page.
- **Bot name**: A title appearing in conversations. Found on the "App Home" page.

#### Responding to a farewell

Chat is a common thing apps do and responding to various types of messages can make conversations more interesting.

Using an editor of choice, open the `app.js` file and add the following message listener after the "hello" handler:

```js
app.message('goodbye', async ({ say }) => {
  const responses = ['Adios', 'Au revoir', 'Farewell'];
  const parting = responses[Math.floor(Math.random() * responses.length)];
  await say(`${parting}!`);
});
```

> Once the file is saved, restart your app to make sure the latest changes are being used then go back to channel and say "goodbye".

Reviewer: ```suggestion
Once the file is saved, restart your app to make sure the latest changes are being used, then go back to channel and say "goodbye".
```

## Next steps {#next-steps}

Congrats once more on getting up and running with this quick start.

> :::note

Reviewer: ```suggestion
:::info[Dive deeper]
```

we don't use "note" because it of the low contrast

Follow along with the steps that went into making this app on the [building an app](/building-an-app) guide for an educational overview.

:::

You can now continue customizing your app with various features to make it right for whatever job's at hand. Here are some ideas about what to explore next:

> * Explore the different events your bot can listen to with the [`event`](/concepts/event-listening) method. All of the [events](https://docs.slack.dev/reference/events) are listed on the documentation site.

Reviewer: ```suggestion
* Explore the different events your bot can listen to with the [`event`](/concepts/event-listening) method. All of the [events](https://docs.slack.dev/reference/events) are listed on the API docs site.
```

* Bolt allows you to call [Web API](/concepts/web-api) methods with the client attached to your app. There are [over 200 methods](https://docs.slack.dev/reference/methods) on our API site.

> * Learn more about the different [token](https://docs.slack.dev/authentication/tokens) types and [authentication](/concepts/authenticating-oauth) setups. Your app might need different tokens depending on the actions you want to perform or for installations to multiple workspaces.

Reviewer: ```suggestion
* Learn more about the different [token types](https://docs.slack.dev/authentication/tokens) and [authentication setups](/concepts/authenticating-oauth). Your app might need different tokens depending on the actions you want to perform or for installations to multiple workspaces.
```

in general, i like to group the entire noun as a link

Author: Heh I shall prefer the recommendation but I like a single words being a link because [underlined spaces](#) make me feel weird 🐍

> * Receive events using HTTP for various deployment methods, such as to the [Heroku](/deployments/heroku) or [AWS Lambda](/deployments/aws-lambda) providers.

Reviewer: ```suggestion
* Receive events using HTTP for various deployment methods, such as deploying to [Heroku](/deployments/heroku) or [AWS Lambda](/deployments/aws-lambda).
```
